feat: add working days component

OussEmaDevCode opened the pull request:
## Description
Added the working days component, connected it to the app via routing and implemented the Required API.

## Related issues and discussion
#162 

## Screenshots, if any
![Capture d’écran (217)](https://user-images.githubusercontent.com/38397893/71545944-b7de0900-2991-11ea-84a5-bf04ddb63f97.png)
![Capture d’écran (218)](https://user-images.githubusercontent.com/38397893/71545955-d47a4100-2991-11ea-8b9b-6c4c079f77b6.png)


## Checklist
Please make sure these boxes are checked before submitting your pull request - thanks!

- [x] If you have multiple commits please combine them into one commit by squashing them.

- [x] Read and understood the contribution guidelines at `web-app/.github/CONTRIBUTING.md`.

OussEmaDevCode commented:
@abhaychawla please have a look !!

abhaychawla commented on src/app/organization/organization-routing.module.ts:
```diff
@@ -18,6 +18,7 @@ import { SmsCampaignsComponent } from './sms-campaigns/sms-campaigns.component';
 import { AdhocQueryComponent } from './adhoc-query/adhoc-query.component';
 import { TellersComponent } from './tellers/tellers.component';
 import { PaymentTypesComponent } from './payment-types/payment-types.component';
+import {WorkingDaysComponent} from './working-days/working-days.component';
```
```suggestion
import { WorkingDaysComponent } from './working-days/working-days.component';
```

abhaychawla commented on src/app/organization/organization.service.ts:
```diff
@@ -82,4 +82,18 @@ export class OrganizationService {
+   * @param {string} workingDays updated working days data.
+   * @returns {Observable<any>}
+   */
+  changeWorkingDays(workingDays: any): Observable<any> {
```
```suggestion
  updateWorkingDays(workingDays: any): Observable<any> {
```

abhaychawla commented on src/app/organization/organization.service.ts:
```diff
@@ -82,4 +82,18 @@ export class OrganizationService {
+  }
+
+  /**
+   * @param {string} workingDays updated working days data.
```
```suggestion
   * @param {string} workingDays Working days data.
```

abhaychawla commented on src/app/organization/working-days/working-days.component.html:
```diff
@@ -0,0 +1,38 @@
+    <div fxLayout="row" fxLayout.lt-md="column">
+      <h3 fxFlex="20%">Working Days</h3>
+      <div fxFlex="80%" fxLayout="column" class="input checkboxes" *ngFor="let day of weekDays; let i = index;">
+        <mat-checkbox [(ngModel)]="weekDays[i].checked">{{day.value}}</mat-checkbox>
```
We are using `reactive forms` in the entire web app not `template-driven forms` (ngModel). Have a look at the other components for reference and update the code accordingly. You can learn more about reactive forms at https://angular.io/guide/reactive-forms

Also, let's keep the container width half like other components which have shorter forms for reference you can see, Accounting -> Closing Entries -> Create

abhaychawla commented on src/app/organization/working-days/working-days.component.html:
```diff
@@ -0,0 +1,38 @@
+    </div>
+
+    <div fxLayout="row" fxLayout.lt-md="column">
+      <h3 fxFlex="20%">Payments due on non working days</h3>
```
To keep consistency with other components keep this as a label instead of the values that you have mentioned as label below: `Payments due on non-working days`

OussEmaDevCode commented:
All requested changes have been made (I have also updated the PR with the new screenshots)

abhaychawla commented on src/app/organization/working-days/working-days.component.html:
```diff
@@ -0,0 +1,50 @@
+          <div fxFlex="60%" class="input">
+            <mat-form-field>
+              <mat-label>Payments due on non working days</mat-label>
+              <mat-select formControlName="selectedOption">
```
Use formControlName as `repaymentsRescheduledType`.

abhaychawla commented on src/app/organization/working-days/working-days.component.html:
```diff
@@ -0,0 +1,50 @@
+            <mat-form-field>
+              <mat-label>Payments due on non working days</mat-label>
+              <mat-select formControlName="selectedOption">
+                <mat-option *ngFor="let option of selectOptions;" [value]="option.id">
```
Use proper name for option and selectOptions.

abhaychawla commented on src/app/organization/working-days/working-days.component.html:
```diff
@@ -0,0 +1,50 @@
+          <div class="input checkboxes" fxFlex="60%">
+            <mat-checkbox formControlName="extendTheTermForDailyRepayments"></mat-checkbox>
+          </div>
+        </div>
```
Remove classes, try to only use flex layout.

abhaychawla commented on src/app/organization/working-days/working-days.component.html:
```diff
@@ -0,0 +1,50 @@
+      </mat-card-content>
+
+      <mat-card-actions fxLayout="row" fxLayout.xs="column" fxLayoutAlign="center" fxLayoutGap="5px">
+        <button mat-raised-button [routerLink]="['..']">Cancel</button>
```
```suggestion
        <button mat-raised-button [routerLink]="['../']">Cancel</button>
```

abhaychawla commented on src/app/organization/working-days/working-days.component.ts:
```diff
@@ -0,0 +1,107 @@
+  /** Working days form. */
+  workingDaysForm: FormGroup;
+  /** Data source for repayment reschedule type select */
+  selectOptions: any;
```
Should be something like `repaymentRescheduleOptionsData`

abhaychawla commented on src/app/organization/working-days/working-days.component.ts:
```diff
@@ -0,0 +1,107 @@
+    for (let i = 0; i < this.weekDays.length; i++) {
+      workingDaysFormData[this.weekDays[i].name] = this.weekDays[i].checked;
+    }
+    this.workingDaysForm = this.formBuilder.group(workingDaysFormData);
```
Use formArray for recurrence in the form group, the format can be modified for the same in submit function. And define repaymentsRescheduledType and extendTermForDailyRepayments in the form group.

abhaychawla commented on src/app/organization/working-days/working-days.component.ts:
```diff
@@ -0,0 +1,107 @@
+   * if successful redirects to organization view.
+   */
+  submit() {
+    const workingDaysFormData = this.workingDaysForm.value;
```
```suggestion
    const workingDays = this.workingDaysForm.value;
```

abhaychawla commented on src/app/organization/working-days/working-days.component.ts:
```diff
@@ -0,0 +1,107 @@
+      // TODO: Update once language and date settings are setup
+      locale: 'en',
+      extendTermForDailyRepayments: workingDaysFormData.extendTheTermForDailyRepayments
+    };
```
Only recurrence parameter should require modification.

abhaychawla commented on src/app/organization/working-days/working-days.component.ts:
```diff
@@ -0,0 +1,107 @@
+      extendTermForDailyRepayments: workingDaysFormData.extendTheTermForDailyRepayments
+    };
+    this.organizationService.updateWorkingDays(workingDays).subscribe(response => {
+      this.router.navigate(['./organization']);
```
Use relative route.

OussEmaDevCode commented:
All requested changes have been made !!! (I have updated the PR with the new screenshots) also apparently whenever I squash a commit git adds it to the commit message that's why Fix appeared thrice but I fixed that problem too.

luckyman20 commented on src/app/organization/organization.service.ts:
```diff
@@ -82,4 +82,18 @@ export class OrganizationService {
   }
 
+  /**
+   * @returns {Observable<any>} Working Days data
```
```suggestion
   * @returns {Observable<any>} Working days data.
```

luckyman20 commented on src/app/organization/working-days/working-days.component.html:
```diff
@@ -0,0 +1,49 @@
+            <h3 fxFlex="40%"> Working Days</h3>
+            <div fxLayout="column" formArrayName="recurrence">
+              <div *ngFor="let day of weekDays; let i = index;">
+                <mat-checkbox [formControlName]="i">{{day.value}}</mat-checkbox>
```
```suggestion
                <mat-checkbox [formControlName]="i"> {{day.value}} </mat-checkbox>
```

OussEmaDevCode commented:
@luckyman20 Done !!!

abhaychawla commented on src/app/organization/working-days/working-days.component.html:
```diff
@@ -0,0 +1,49 @@
+          <mat-form-field>
+            <mat-label>Payments due on non working days</mat-label>
+            <mat-select formControlName="repaymentRescheduleType">
+              <mat-option *ngFor="let option of repaymentRescheduleOptionsData;" [value]="option.id">
```
```suggestion
              <mat-option *ngFor="let repaymentRescheduleType of repaymentRescheduleTypeData" [value]="repaymentRescheduleType.id">
```

abhaychawla commented on src/app/organization/working-days/working-days.component.html:
```diff
@@ -0,0 +1,49 @@
+            <mat-label>Payments due on non working days</mat-label>
+            <mat-select formControlName="repaymentRescheduleType">
+              <mat-option *ngFor="let option of repaymentRescheduleOptionsData;" [value]="option.id">
+                {{option.value}}
```
```suggestion
                {{ repaymentRescheduleType.value }}
```

abhaychawla commented on src/app/organization/working-days/working-days.component.html:
```diff
@@ -0,0 +1,49 @@
+            <div fxFlex="60%">
+              <mat-checkbox formControlName="extendTermForDailyRepayments"></mat-checkbox>
+            </div>
+          </div>
```
```
          <mat-checkbox labelPosition="before" formControlName="extendTermForDailyRepayments">
            Extend the term for loans following a daily repayment schedule
          </mat-checkbox>
```

abhaychawla commented on src/app/organization/working-days/working-days.component.html:
```diff
@@ -0,0 +1,49 @@
+
+      <mat-card-actions fxLayout="row" fxLayout.xs="column" fxLayoutAlign="center" fxLayoutGap="5px">
+        <button mat-raised-button [routerLink]="['../']">Cancel</button>
+        <button mat-raised-button color="primary">Submit</button>
```
The button should be disabled until any change is made (as in other edit components of the web app)

abhaychawla commented on src/app/organization/working-days/working-days.component.ts:
```diff
@@ -0,0 +1,113 @@
+import { Component, OnInit } from '@angular/core';
+import { FormGroup, FormBuilder } from '@angular/forms';
+import { Router, ActivatedRoute } from '@angular/router';
+import { OrganizationService } from '../organization.service';
```
Not under the correct comment.

abhaychawla commented on src/app/organization/working-days/working-days.component.ts:
```diff
@@ -0,0 +1,113 @@
+import { OrganizationService } from '../organization.service';
+
+/**
+ * Working-days component.
```
```suggestion
 * Working days component.
```

abhaychawla commented on src/app/organization/working-days/working-days.component.ts:
```diff
@@ -0,0 +1,113 @@
+  }
+
+  /**
+   * Edits the current working days configuration,
```
```suggestion
   * Submits the working days form and updates working days configuration,
```

abhaychawla commented on src/app/organization/working-days/working-days.component.ts:
```diff
@@ -0,0 +1,113 @@
+    {name: 'FR', value: 'Friday', checked: false},
+    {name: 'SA', value: 'Saturday', checked: false},
+    {name: 'SU', value: 'Sunday', checked: false}
+  ];
```
```
  /** Week days */
  weekDays = [
    { name: 'Monday', value: 'MO', checked: false },
    { name: 'Tuseday', value: 'TU', checked: false },
    { name: 'Wednesday', value: 'WE', checked: false },
    { name: 'Thursday', value: 'TH', checked: false },
    { name: 'Friday', value: 'FR', checked: false },
    { name: 'Saturday', value: 'SA', checked: false },
    { name: 'Sunday', value: 'SU', checked: false }
  ];
```

abhaychawla commented on src/app/organization/working-days/working-days.component.ts:
```diff
@@ -0,0 +1,113 @@
+  /** Working days form. */
+  workingDaysForm: FormGroup;
+  /** Data source for repayment reschedule type select */
+  repaymentRescheduleOptionsData: any;
```
```  
  /** Repayment schedule type data. */
  repaymentRescheduleTypeData: any;
```

abhaychawla commented on src/app/organization/working-days/working-days.component.ts:
```diff
@@ -0,0 +1,113 @@
+
+  /**
+   * Retrieves the working days data from `resolve`.
+   * @param {formBuilder} formBuilder form builder.
```
```suggestion
   * @param {formBuilder} formBuilder Form Builder.
```

abhaychawla commented on src/app/organization/working-days/working-days.component.ts:
```diff
@@ -0,0 +1,113 @@
+   * Retrieves the working days data from `resolve`.
+   * @param {formBuilder} formBuilder form builder.
+   * @param {ActivatedRoute} route Activated Route.
+   * @param {organizationService} organizationService organization service.
```
```suggestion
   * @param {organizationService} organizationService Organization Service.
```

abhaychawla commented on src/app/organization/working-days/working-days.component.ts:
```diff
@@ -0,0 +1,113 @@
+   * @param {formBuilder} formBuilder form builder.
+   * @param {ActivatedRoute} route Activated Route.
+   * @param {organizationService} organizationService organization service.
+   * @param {Router} router Router for navigation
```
```suggestion
   * @param {Router} router Router for navigation.
```

abhaychawla commented on src/app/organization/working-days/working-days.component.ts:
```diff
@@ -0,0 +1,113 @@
+   * Creates the working days form.
+   */
+  createWorkingDaysForm() {
+    this.setWeekWorkingDays();
```
```suggestion
    this.setWeeklyWorkingDays();
```

abhaychawla commented on src/app/organization/working-days/working-days.component.ts:
```diff
@@ -0,0 +1,113 @@
+  }
+
+  /**
+   * Initialize days checkboxes
```
```suggestion
   * Sets weekly working days.
```

abhaychawla commented on src/app/organization/working-days/working-days.component.ts:
```diff
@@ -0,0 +1,113 @@
+  /**
+   * Initialize days checkboxes
+   */
+  setWeekWorkingDays() {
```
```suggestion
  setWeeklyWorkingDays() {
```

abhaychawla commented on src/app/organization/working-days/working-days.component.ts:
```diff
@@ -0,0 +1,113 @@
+  setWeekWorkingDays() {
+    const days = this.workingDaysData.recurrence.replace('FREQ=WEEKLY;INTERVAL=1;BYDAY=', '');
+    for (let i = 0; i < this.weekDays.length; i++) {
+      this.weekDays[i].checked = days.includes(this.weekDays[i].name);
```
```suggestion
      this.weekDays[i].checked = days.includes(this.weekDays[i].value);
```

abhaychawla commented on src/app/organization/working-days/working-days.component.ts:
```diff
@@ -0,0 +1,113 @@
+      day.setValue(this.weekDays[i].checked);
+      recurrence.push(day);
+    }
+    return recurrence;
```
```
return this.weekDays.map(weekDay => new FormControl(weekDay.checked));
```

This can be made simpler.

abhaychawla commented on src/app/organization/working-days/working-days.component.ts:
```diff
@@ -0,0 +1,113 @@
+   */
+  submit() {
+    const workingDays = this.workingDaysForm.value;
+    let recurrence = 'FREQ=WEEKLY;INTERVAL=1;BYDAY=';
```
Create a constant above the class declaration and use it instead. (you can use it here as well as above in the replace function instead of writing the same string again)

abhaychawla commented on src/app/organization/working-days/working-days.component.ts:
```diff
@@ -0,0 +1,113 @@
+    let recurrence = 'FREQ=WEEKLY;INTERVAL=1;BYDAY=';
+    for (let i = 0; i < this.weekDays.length; i++) {
+      if (workingDays.recurrence[i]) {
+        recurrence = recurrence + this.weekDays[i].name + ',';
```
```suggestion
        recurrence = recurrence + this.weekDays[i].value + ',';
```

OussEmaDevCode commented:
@abhaychawla I've done the changes you requested and as always updated the PR message with the new UI, I have also performed some small changes to the component so that it's consistent with the other components.

abhaychawla commented on src/app/organization/working-days/working-days.component.ts:
```diff
@@ -0,0 +1,119 @@
+  /**
+   * @returns {FormArray} recurrence form array.
+   */
+  getRecurrenceFormArray() {
```
```suggestion
  get recurrence(): FormArray {
```

abhaychawla commented on src/app/organization/working-days/working-days.component.html:
```diff
@@ -0,0 +1,50 @@
+          <div fxLayout="row" fxLayout.lt-md="column">
+            <mat-label fxFlex="50%" class="mat-h4">Working Days</mat-label>
+            <div fxFlex="50%" fxLayout="column" formArrayName="recurrence">
+              <div *ngFor="let day of getRecurrenceFormArray().controls; let i = index;">
```
```suggestion
              <div *ngFor="let day of recurrence.controls; let i = index;">
```

abhaychawla commented on src/app/organization/working-days/working-days.component.html:
```diff
@@ -0,0 +1,50 @@
+            <mat-label>Payments due on non working days</mat-label>
+            <mat-select formControlName="repaymentRescheduleType">
+              <mat-option *ngFor="let repaymentRescheduleType of repaymentRescheduleTypeData;" [value]="repaymentRescheduleType.id">
+                {{repaymentRescheduleType.value}}
```
```suggestion
                {{ repaymentRescheduleType.value }}
```

abhaychawla commented on src/app/organization/working-days/working-days.component.html:
```diff
@@ -0,0 +1,50 @@
+          <mat-form-field>
+            <mat-label>Payments due on non working days</mat-label>
+            <mat-select formControlName="repaymentRescheduleType">
+              <mat-option *ngFor="let repaymentRescheduleType of repaymentRescheduleTypeData;" [value]="repaymentRescheduleType.id">
```
```suggestion
              <mat-option *ngFor="let repaymentRescheduleType of repaymentRescheduleTypeData" [value]="repaymentRescheduleType.id">
```

abhaychawla commented on src/app/organization/working-days/working-days.component.html:
```diff
@@ -0,0 +1,50 @@
+            </mat-select>
+          </mat-form-field>
+
+          <div fxLayout="row" fxLayout.lt-md="column">
```
```suggestion
```
Remove this.

abhaychawla commented on src/app/organization/working-days/working-days.component.html:
```diff
@@ -0,0 +1,50 @@
+            <mat-checkbox labelPosition="before" formControlName="extendTermForDailyRepayments">
+              Extend the term for loans following a daily repayment schedule
+            </mat-checkbox>
+          </div>
```
```suggestion
```
Remove this.

abhaychawla commented on src/app/organization/working-days/working-days.component.html:
```diff
@@ -0,0 +1,50 @@
+
+      <mat-card-actions fxLayout="row" fxLayout.xs="column" fxLayoutAlign="center" fxLayoutGap="5px">
+        <button mat-raised-button [routerLink]="['../']">Cancel</button>
+        <button mat-raised-button color="primary" [disabled]="!workingDaysForm.dirty">Submit</button>
```
```suggestion
        <button mat-raised-button color="primary" [disabled]="workingDaysForm.pristine">Submit</button>
```
Pristine property is being used throughout the web app. So we will use that.

abhaychawla commented on src/app/organization/working-days/working-days.component.ts:
```diff
@@ -0,0 +1,119 @@
+  workingDaysData: any;
+  /** Week days */
+  weekDays = [
+    {name: 'Monday', value: 'MO', checked: false},
```
```suggestion
    { name: 'Monday', value: 'MO', checked: false },
```
Please add one space between the text and the brackets everywhere.

OussEmaDevCode commented:
@abhaychawla Required changes have been made !

punwai commented:
I recommend that you rebase the recent changes to your branch rather than merging, as it will be harder for the reviewers and will add an extra commit apart from yours to the master branch.

OussEmaDevCode commented:
I haven't found a way to revert the merge so I think I'll create a new cleaner PR
